### PR TITLE
DDBEASY-116 - Refactored the availability and reservable logic.

### PR DESCRIPTION
**Pull request by guddo**
http://platform.dandigbib.org/issues/618

### CODE REVIEW

File: ding_availability.make [hunk 1]
```diff
@@ -11,5 +11,5 @@ projects[virtual_field][version] = "1.2"
 projects[ding_provider][type] = "module"
 projects[ding_provider][download][type] = "git"
 projects[ding_provider][download][url] = "git@github.com:ding2/ding_provider.git"
-projects[ding_provider][download][branch] = "master"
+projects[ding_provider][download][tag] = "v2.0.0"
```
Comment (kasperg) [inline ding_availability.make]: This change should not be a part of the pull request.

File: ding_availability.module [hunk 2]
```diff
@@ -151,6 +151,7 @@ function ding_availability_items($provider_ids) {
  */
 function ding_availability_holdings($provider_ids) {
   $items = ding_provider_invoke('availability', 'holdings', $provider_ids);
+  $items = array_filter($items, function($v) { return $v['availability_information']; });
```
Comment (kasperg) [inline ding_availability.module]: Please add a comment explaining why this is necessary. The `availability_information` entry does not seem to be used further down the line.

Since the `array_filter` callback expects a boolean I would appreciate wrapping `$v['availability_information'];` in `empty()`. This should also prevent warnings if the entry is missing.

File: ding_availability.module [hunk 3]
```diff
@@ -184,7 +185,25 @@ function ding_availability_holdings($provider_ids) {
...
+          $copies++;
+          if (empty($closest_loan) || $closest_loan < $next_loan_time) {
+            $closest_loan = $next_loan_time;
+          }
+        }
+
+        if ($holding['status'] == 'inAcquisition') {
+          $in_acquisition++;
+        }
+
```
Comment (kasperg) [inline ding_availability.module]: Is this change intentionally part of this pull request?

Extracting `closest_loan` and `in_acquisition` does not seem to be related to [issue 618](http://platform.dandigbib.org/issues/618). I would also think that supporting this would require changes to the provider implementations.

File: ding_availability.module [hunk 4]
```diff
@@ -198,6 +217,8 @@ function ding_availability_holdings($provider_ids) {
         ),
         'total_count' => $item['total_count'],
         'reserved_count' => $item['reserved_count'],
+        'closest_loan' => (count($item['holdings']) == $copies) ? ding_availability_holding_format_time($closest_loan) : NULL,
+        'in_acquisition' => $in_acquisition,
```
Comment (kasperg) [inline ding_availability.module]: See above.

File: ding_availability.module [hunk 5]
```diff
@@ -206,6 +227,39 @@ function ding_availability_holdings($provider_ids) {
...
+    $label = 'day';
+  }
+
+  if ($days > 1) {
+    $label = $plurals[$label];
+  }
+
+  $text = t('The waiting time is currently: <strong>@days @label</strong>', array('@days' => $days, '@label' => $label));
+
+  return $text;
```
Comment (kasperg) [inline ding_availability.module]: See above.

Also: How about using [`format_interval`](https://api.drupal.org/api/drupal/includes%21common.inc/function/format_interval/7) for this?

**Comment by guddo** [comment]
The only change related to [issue 618](http://platform.dandigbib.org/issues/618) is what has been done in this commit: https://github.com/easyddb/ding_availability/commit/dc1fb96e3a0ee2ee0a36fdd41d146f678313f57b.

This branch is branched out from the previous PRs which means #9, #10, #11, #12 and #13 is a part of this PR.

File: js/ding_availability.js [hunk 6]
```diff
@@ -130,6 +130,9 @@
           }
         }
       }
+      else {
+        reserver_btn.addClass('not-reservable');
+      }
```
Comment (kasperg) [inline js/ding_availability.js]: Why add the `not-reservable` class to an element which we do not know the status for?

If we are to add a class I would expect it that we should add it to the element as well as in the other branches of the code.

File: js/ding_availability.js [hunk 7]
```diff
@@ -148,6 +151,13 @@
     if (Drupal.DADB[entity_id] && (Drupal.DADB[entity_id]['holdings'])) {
       // Show status for material.
       $('#' + id).html(Drupal.DADB[entity_id].html);
+      // Don't show queue time if item not reservable.
+      if (Drupal.DADB[entity_id].reservable === false) {
+        $('#' + id + ' span.in-queue').hide();
+      }
+    }
+    else {
+      $('div.group-holdings-available').parent().parent().remove();
```
Comment (kasperg) [inline js/ding_availability.js]: See comment above regarding queue time.

File: js/ding_availability_labels.js [hunk 8]
```diff
@@ -182,16 +173,32 @@
...
+            }
+            else if (i === 'reservable') {
+              class_name = 'not-' + i;
+            }
+          }
+
+          element.addClass(class_name);
+          if (btn.length) {
+            btn.addClass(class_name);
+          }
```
Comment (kasperg) [inline js/ding_availability_labels.js]: Well done!

Is this none usable in [`ding_availability.js`](https://github.com/ding2/ding_availability/pull/15/files#diff-c7980ed510e19f8dc6f2b06c892ee392R101)? It seems as if it addresses the exact same things.

File: templates/ding-holdings.tpl.php [hunk 9]
```diff
@@ -9,8 +9,19 @@
...
+}
+
 ?>
-<p><?php print "$total_text $reserved_text"; ?></p>
+<p>
+  <span><?php print $total_text . "<span class='in-queue'>" . $reserved_text . "</span>"; ?></span>
+</p>
+<p>
+  <?php print $closest_loan; ?>
+</p>
```
Comment (kasperg) [inline templates/ding-holdings.tpl.php]: See comment above regarding acquisition and queue time.

**Comment by kasperg** [comment]
Review complete. Well done regarding the refactoring of the availability logic.

This pull request contains more changes than has been approved for the next release and thus I cannot approve it without further consent from the product owners.

**Comment by kasperg** [comment]
Also: Please rebase this pull request. It is no longer a trivial merge.

**Comment by guddo** [comment]
https://github.com/ding2/ding_availability/pull/16

**Comment by guddo** [comment]
*Note to POs: There has been no resource allocation by you for issues like this. Please don't expect us to contribute based on the status of current ding2 code nor to modify already existing PRs, if they are dependant from other PRs.*
